pravega/src/main/scala/akka/stream/alpakka/pravega/impl/PravegaTableSource.scala

```diff
@@ -0,0 +1,164 @@
+import io.pravega.client.KeyValueTableFactory
+import io.pravega.client.tables.{IteratorItem, KeyValueTable, KeyValueTableClientConfiguration, TableEntry}
+
+import scala.collection.mutable
+
+import java.util.concurrent.Semaphore
+import io.pravega.common.util.AsyncIterator
+@InternalApi private final class PravegaTableSourceStageLogic[K, V](
+    shape: SourceShape[(K, V)],
+    val scope: String,
+    tableName: String,
+    keyFamily: String,
+    tableSettings: TableSettings[K, V],
+    startupPromise: Promise[Done]
+) extends GraphStageLogic(shape)
+    with StageLogging {
+
+  override protected def logSource = classOf[PravegaTableSourceStageLogic[K, V]]
+
+  private def out = shape.out
+
+  private var keyValueTableFactory: KeyValueTableFactory = _
+  private var table: KeyValueTable[K, V] = _
+
+  private val queue = mutable.Queue.empty[(K, V)]
+
+  private val semaphore = new Semaphore(tableSettings.maximumInflightMessages)
+
+  private var closing = false
+
+  val logThat: AsyncCallback[String] = getAsyncCallback { message =>
+    log.info(message)
+  }
+
+  private def pushElement(out: Outlet[(K, V)], element: (K, V)) = {
+    push(out, element)
+    semaphore.release()
+  }
+
+  val onElement: AsyncCallback[(K, V)] = getAsyncCallback[(K, V)] { element =>
+    if (isAvailable(out) && queue.isEmpty)
+      pushElement(out, element)
+    else
+      queue.enqueue(element)
+
+  }
+
+  val onFinish: AsyncCallback[Unit] = getAsyncCallback[Unit] { _ =>
+    closing = true
+    if (queue.isEmpty)
+      completeStage()
+
+  }
+
+  setHandler(
+    out,
+    new OutHandler {
+      override def onPull(): Unit = {
+        if (!queue.isEmpty)
+          pushElement(out, queue.dequeue())
+        if (closing && queue.isEmpty)
+          completeStage()
+      }
+    }
+  )
+
+  def nextIteration(iterator: AsyncIterator[IteratorItem[TableEntry[K, V]]]): Unit =
+    iterator.getNext
+      .thenAccept(new Consumer[IteratorItem[TableEntry[K, V]]] {
+        override def accept(iteratorItem: IteratorItem[TableEntry[K, V]]): Unit = {
+          if (iteratorItem == null) {
+            onFinish.invoke(())
+          } else {
+            iteratorItem.getItems.stream().forEach { tableEntry =>
+              semaphore.acquire()
+              onElement.invoke((tableEntry.getKey.getKey, tableEntry.getValue))
+            }
+            nextIteration(iterator)
+          }
+        }
+      })
+
+  override def preStart(): Unit = {
+    log.info("Start consuming {} by {} ...", tableName, tableSettings.maximumInflightMessages)
+    try {
+
+      val kvtClientConfig = KeyValueTableClientConfiguration.builder().build()
+      keyValueTableFactory = KeyValueTableFactory
+        .withScope(scope, tableSettings.clientConfig)
+
+      table = keyValueTableFactory
+        .forKeyValueTable(tableName, tableSettings.keySerializer, tableSettings.valueSerializer, kvtClientConfig)
+
+      //TODO Applied * 10, but no idea :-/
```

Comment:
Is there a convention from the Pravega community we can follow here?  Maybe this should be separate configuration?

Reply:
Yes I will put this in settings.